harness/determined/exec/harness.py

```diff
@@ -82,69 +83,75 @@ def maybe_load_checkpoint(
+        print("EventSender is ", type(event_trail))
 
-        # Create the storage manager. This is used to download the initial checkpoint here in
-        # build_training_pipeline and also used by the workload manager to create and store
-        # checkpoints during training.
-        storage_mgr = storage.build(
-            env.experiment_config["checkpoint_storage"],
-            container_path=constants.SHARED_FS_CONTAINER_PATH,
-        )
-
-        [tensorboard_mgr, tensorboard_writer] = load.prepare_tensorboard(
-            env, constants.SHARED_FS_CONTAINER_PATH
-        )
-
-        # Create the workload manager. The workload manager will receive workloads from the
-        # socket_mgr, and augment them with some additional arguments. Additionally, the
-        # workload manager is responsible for some generic workload hooks for things like timing
-        # workloads, preparing checkpoints, and uploading completed checkpoints.  Finally, the
-        # workload manager does some sanity checks on response messages that originate from the
-        # trial.
+        # Create the socket manager. The socket manager will connect to the master and read messages
+        # until it receives the rendezvous_info.
         #
-        # TODO(ryan): Refactor WorkloadManager into separate layers that do each separate task.
-        workload_mgr = layers.build_workload_manager(
-            env,
-            iter(socket_mgr),
-            socket_mgr.get_rendezvous_info(),
-            storage_mgr,
-            tensorboard_mgr,
-            tensorboard_writer,
-        )
-
-        hvd_config = horovod.HorovodContext.from_configs(
-            env.experiment_config, socket_mgr.get_rendezvous_info(), env.hparams
-        )
-        logging.info(f"Horovod config: {hvd_config.__dict__}.")
-
-        # Load the checkpoint, if necessary. Any possible sinks to this pipeline will need access
-        # to this checkpoint.
-        with maybe_load_checkpoint(storage_mgr, env.latest_checkpoint) as load_path:
-
-            # Horovod distributed training is done inside subprocesses.
-            if hvd_config.use:
-                subproc = layers.SubprocessLauncher(
-                    env, iter(workload_mgr), load_path, socket_mgr.get_rendezvous_info(), hvd_config
-                )
-                subproc.run()
-            else:
-                if env.experiment_config.debug_enabled():
-                    faulthandler.dump_traceback_later(30, repeat=True)
-
-                with det._catch_sys_exit():
-                    controller = load.prepare_controller(
-                        env,
-                        iter(workload_mgr),
-                        load_path,
-                        socket_mgr.get_rendezvous_info(),
-                        hvd_config,
+        # TODO(ryan): Pull profiler hooks out of SocketManager and into their own layer.
+        with layers.SocketManager(env) as socket_mgr:
```

Comment:
@rb-determined-ai I think `SocketManager` should also be a singleton rather than a context manager.

Reply:
SocketManager also needs cleanup and makes a lot of sense as a context manager.